http/digest/src/main/java/org/wildfly/security/http/digest/PersistentNonceManager.java

```diff
@@ -0,0 +1,228 @@
+     * @return a new encoded nonce to send to the client.
+     */
+    String generateNonce(byte[] salt) {
+        return NonceManagerUtils.generateNonce(salt, algorithm, nonceCounter, privateKey );
+    }
+
+    /**
+     * Attempt to use the supplied nonce.
+     *
+     * A nonce might not be usable for a couple of different reasons: -
+     *
+     * <ul>
+     *     <li>It was created too far in the past.
+     *     <li>Validation of the signature fails.
+     *     <li>The nonce has been used previously and re-use is disabled.
+     * </ul>
+     *
+     * @param nonce the nonce supplied by the client.
+     * @param nonceCount the nonce count, or -1 if not present
+     * @return {@code true} if the nonce can be used, {@code false} otherwise.
+     * @throws AuthenticationMechanismException
+     */
+    boolean useNonce(String nonce, int nonceCount) throws AuthenticationMechanismException {
+        return useNonce(nonce, null, nonceCount);
+    }
+
+    /**
+     * Attempt to use the supplied nonce.
+     *
+     * A nonce might not be usable for a couple of different reasons: -
+     *
+     * <ul>
+     *     <li>It was created too far in the past.
+     *     <li>Validation of the signature fails.
+     *     <li>The nonce has been used previously and re-use is disabled.
+     * </ul>
+     *
+     * @param nonce the nonce supplied by the client.
+     * @param salt additional data to use when creating the overall signature for the nonce.
+     * @return {@code true} if the nonce can be used, {@code false} otherwise.
+     * @throws AuthenticationMechanismException
+     */
+    boolean useNonce(final String nonce, byte[] salt, int nonceCount) throws AuthenticationMechanismException {
+        return NonceManagerUtils.useNonce(nonce,  salt,  nonceCount, algorithm, privateKey, usedNonces,  validityPeriodNano,  executor, singleUse, nonceSessionTime);
+    }
+
+    ScheduledExecutorService getExecutor() {
+        return executor;
+    }
+
+    AtomicInteger getNonceCounter() {
+        return nonceCounter;
+    }
+
+    Map<String, NonceManagerUtils.NonceState> getUsedNonces() {
+        return usedNonces;
+    }
+
+    byte[] getPrivateKey() {
+        return privateKey;
+    }
+    long getValidityPeriodNano() {
+        return validityPeriodNano;
+    }
+
+    long getNonceSessionTime() {
+        return nonceSessionTime;
+    }
+
+    String getAlgorithm() {
+        return algorithm;
+    }
+
+    void setAlgorithm(String algorithm) {
+        this.algorithm = algorithm;
+    }
+
+    boolean isSingleUse() {
+        return singleUse;
+    }
+
+    /**
+     * Update attributes from provided PersistentNonceManager to this instance
+     *
+     * @param persistentNonceManager PersistentNonceManager instance to update the attributes from
+     */
+    void refreshInfoFromSessionNonceManager(PersistentNonceManager persistentNonceManager) {
```

Comment:
Just so I understand better, would you be able to provide some details on why this is needed? Wondering if it's possible to make use of the nonce manager from the session directly rather than updating the attributes of another instance.

Reply:
@fjuma The` nonceManager` in [DigestAuthenticationmechanism](https://github.com/wildfly-security/wildfly-elytron/blob/2.x/http/digest/src/main/java/org/wildfly/security/http/digest/DigestAuthenticationMechanism.java#L86), where this method is being used, is a final variable. I need to  be retrieving the nonce manager with each request, but I cannot assign to it. So I am assigning to the attributes instead. Should I change the nonceManager to be non final so I can assign to it with every request?

Reply:
I removed the `final` from the nonceManager and removed this method. It is now being assigned directly instead of updating specific attributes